### PR TITLE
[Model] RGCN with new heterograph API

### PR DESCRIPTION
**Pull request by isratnisa**
## Description
The new heterogeneous graph APIs, .e.g, `update_all`, `apply_edges` invoke one python call to handle all the relation types of a heterogeneous graph and performs one-level message aggregation. Previously, either separate python kernels were invoked for each relation type, or converts the graph to a homogeneous one which have one relation type and calls one python kernel. This PR implements RGCN-hetero with the new APIs and achieves `6x` speedup on `aifb` and `mutag` datasets compared to the old RGCN-hetero code. The corresponding NN modules should be changed to allow the new API. For debugging purpose we modified only the example codes. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/rgcn-hetero/model.py [hunk 1]
```diff
@@ -120,6 +121,125 @@ def _apply(ntype, h):
...
+        self.out_feat = out_feat
+        self.rel_names = rel_names
+        self.num_bases = num_bases
+        self.bias = bias
+        self.activation = activation
+        self.self_loop = self_loop
+
+        self.use_weight = weight
+        self.use_basis = num_bases < len(self.rel_names) and weight
+        if self.use_weight:
```
Comment (zheng-da) [inline examples/pytorch/rgcn-hetero/model.py:170]: i'm just curious, why do we have this option? if not using weight, is it a graphsage layer?

Comment (zheng-da) [reply examples/pytorch/rgcn-hetero/model.py:170]: it's not even graphsage. there will be no parameters in the model.

File: examples/pytorch/rgcn-hetero/model.py [hunk 2]
```diff
@@ -120,6 +121,125 @@ def _apply(ntype, h):
...
+            weight = self.basis() if self.use_basis else self.weight
+            wdict = {self.rel_names[i] : {'weight' : w.squeeze(0)}
+                     for i, w in enumerate(th.split(weight, 1, dim=0))}
+        else:
+            wdict = {}
+
+        if g.is_block:
+            inputs_src = inputs
+            inputs_dst = {k: v[:g.number_of_dst_nodes(k)] for k, v in inputs.items()}
```
Comment (zheng-da) [inline examples/pytorch/rgcn-hetero/model.py:215]: is this part of code tested? i suppose you only have full-graph training?

File: examples/pytorch/rgcn-hetero/model.py [hunk 3]
```diff
@@ -253,3 +373,98 @@ def inference(self, g, batch_size, device, num_workers, x=None):
...
+
+        ***NOTE***
+        For node classification, the model is trained to predict on only one node type's
+        label.  Therefore, only that type's final representation is meaningful.
+        """
+
+        if x is None:
+            x = self.embed_layer()
+
+        for l, layer in enumerate(self.layers):
```
Comment (zheng-da) [inline examples/pytorch/rgcn-hetero/model.py:443]: it seems you train the model with full graph but perform inference with mini-batch. we normally don't do it.

**Comment by zheng-da** [comment]
will you implement a mini-batch version as well?

File: examples/pytorch/rgcn-hetero/model.py (hunk 2 above)
Comment (isratnisa) [reply examples/pytorch/rgcn-hetero/model.py:215]: Removed for now. Will add back to support minibatch version.

File: examples/pytorch/rgcn-hetero/model.py (hunk 3 above)
Comment (isratnisa) [reply examples/pytorch/rgcn-hetero/model.py:440]: I see. Apparently, the original RGCN-hetero example with `heteroConv` also using the mini-batch based inferencing. Both cases should be changed?

Comment (zheng-da) [reply examples/pytorch/rgcn-hetero/model.py:440]: this implementation is very weird.

**Review by zheng-da — APPROVED** [review]

**Pull request merged**